pkg/networkserver/adr.go

```diff
@@ -85,6 +85,48 @@ func deviceADRMargin(dev *ttnpb.EndDevice, defaults ttnpb.MACSettings) float32 {
 	return DefaultADRMargin
 }
 
+func lossRate(nbTrans uint32, ups ...*ttnpb.UplinkMessage) float32 {
+	if len(ups) < 2 {
+		return 0
+	}
+
+	min := ups[0].Payload.GetMACPayload().FHDR.FCnt
+	max := ups[len(ups)-1].Payload.GetMACPayload().FHDR.FCnt
+
+	fCnt := min + 1
+	var fCntTrans uint32
+	var lost uint32
+	n := uint32(len(ups))
+
+loop:
+	for i := 0; i < len(ups); i++ {
+		switch v := ups[i].Payload.GetMACPayload().FHDR.FCnt; v {
+		case min:
+			continue
+
+		case fCnt:
+			fCntTrans++
+
+		default:
+			if fCntTrans < nbTrans {
+				d := nbTrans - fCntTrans
+				lost += d
+				n += d
+			}
+			d := (v - fCnt - 1) * nbTrans
```

Comment:
This goes wrong on an FCnt reset, right? Or do we clear recent uplinks on reset?

Reply:
This will be done in #505